### PR TITLE
Use `struct` instead of `named_struct` when there are no aliases

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/9839

## Rationale for this change

The names of the columns created by callling `struct` changed in https://github.com/apache/arrow-datafusion/pull/9743

### Before https://github.com/apache/arrow-datafusion/pull/9743

```sql
select struct(a, b) from t;
+-----------------+
| struct(t.a,t.b) |
+-----------------+
| {c0: 1, c1: 2}  |
| {c0: 3, c1: 4}  |
+-----------------+
```

### After https://github.com/apache/arrow-datafusion/pull/9743

(note how the column name is different)

```sql
select struct(a as field_a, b) from t;
+--------------------------------------------------+
| named_struct(Utf8("field_a"),t.a,Utf8("c1"),t.b) |
+--------------------------------------------------+
| {field_a: 1, c1: 2}                              |
| {field_a: 3, c1: 4}                              |
+--------------------------------------------------+
```

This change caused https://github.com/apache/arrow-datafusion/issues/9891 as well as removed test coverage for `struct` 

## What changes are included in this PR?

1. Use `struct` when there are no named fields
2. reverts the test change in https://github.com/apache/arrow-datafusion/pull/9894

## Are these changes tested?
Yes, by existing tests

## Are there any user-facing changes?

The column names are now the same as they were prior to https://github.com/apache/arrow-datafusion/pull/9894

### CODE REVIEW

File: datafusion/sqllogictest/test_files/expr.slt [hunk 1]
```diff
@@ -2288,39 +2288,39 @@ select struct(time,load1,load2,host) from t1;
 
 # can have an aggregate function with an inner coalesce
 query TR
-select t2.info['c3'] as host, sum(coalesce(t2.info)['c1']) from (select struct(time,load1,load2,host) as info from t1) t2 where t2.info['c3'] IS NOT NULL group by t2.info['c3'] order by host;
+select t2."struct(t1.time,t1.load1,t1.load2,t1.host)"['c3'] as host, sum(coalesce(t2."struct(t1.time,t1.load1,t1.load2,t1.host)")['c1']) from (select struct(time,load1,load2,host) from t1) t2 where t2."struct(t1.time,t1.load1,t1.load2,t1.host)"['c3'] IS NOT NULL group by t2."struct(t1.time,t1.load1,t1.load2,t1.host)"['c3'] order by host;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/expr.slt:2291]: This reverts the change from https://github.com/apache/arrow-datafusion/pull/9894

File: datafusion/sql/src/expr/mod.rs [hunk 2]
```diff
@@ -603,6 +604,25 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         if !fields.is_empty() {
             return not_impl_err!("Struct fields are not supported yet");
         }
+
+        if values
+            .iter()
+            .any(|value| matches!(value, SQLExpr::Named { .. }))
+        {
+            self.create_named_struct(values, input_schema, planner_context)
```
Comment (yyy1000) [inline datafusion/sql/src/expr/mod.rs:612]: I think the values in `create_named_struct` will be `SQLExpr::Named` in this PR, so maybe we can also make a change here?
https://github.com/apache/arrow-datafusion/blob/d8d521ac8b90002fa0ba1f91456051a9775ae193/datafusion/sql/src/expr/mod.rs#L610-L631

**Review by comphead — APPROVED** [review]
thanks @alamb 1 improvement I can see here is we over iterating through values.
I mean 1 iteration is do identify named/not_named and when creating struct/named_struct its another one. prob it can be optimized

File: datafusion/sql/src/expr/mod.rs (hunk 2 above)
Comment (yyy1000) [reply datafusion/sql/src/expr/mod.rs:612]: Oh, ignore this. One case is only some values are Named so no change is needed.

**Comment by alamb** [comment]
> thanks @alamb 1 improvement I can see here is we over iterating through values. I mean 1 iteration is do identify named/not_named and when creating struct/named_struct its another one. prob it can be optimized

I agree it does two iterations over the value. I think the current approach basically requires two iterations because if we are doing `named_struct` we need to add names for each column that don't have a name.

So in other words, `struct(1, 2, 3 as other_col)` needs to become `struct(1 as c1, 2 as c2, 3 as other_col)` but we don't know  until we see `3 as other_col` that we need to add in `c1` and `c2`) 🤔 

I would like to spend my optimization budget on bigger things initially if that is ok (specifically https://github.com/apache/arrow-datafusion/issues/9637)

**Comment by alamb** [comment]
Thanks again for the reviews 🙏

**Pull request merged**